### PR TITLE
IGNITE-18073 Add indexes to API for full rebalance without losing user data in MvPartitionStorage on receiver

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-18073

### CODE REVIEW

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/MvTableStorage.java [hunk 1]
```diff
@@ -164,43 +166,47 @@ default IndexStorage getOrCreateIndex(int partitionId, UUID indexId) {
...
-     * <p>This method must be called before every full rebalance of the partition storage, so that in case of errors or cancellation of the
-     * full rebalance, we can restore the partition storage from the backup.
+     * <p>This method must be called before every full rebalance, so that in case of errors or cancellation of the full rebalance, we can
+     * restore the partition storage and its index storages from the backup.
      *
      * <p>Full rebalance will be completed when one of the methods is called:
      * <ol>
      *     <li>{@link #abortRebalanceMvPartition(int)} - in case of a full rebalance cancellation or failure, so that we can
-     *     restore the partition storage from a backup;</li>
+     *     restore thepartition storage and its index storages from a backup;</li>
```
Comment (ibessonov) [inline modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/MvTableStorage.java:178]:
```suggestion
     *     restore the partition storage and its index storages from a backup;</li>
```

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/MvTableStorage.java [hunk 2]
```diff
@@ -116,16 +116,18 @@ default IndexStorage getOrCreateIndex(int partitionId, UUID indexId) {
...
      * Destroys the index under the given name and all data in it.
      *
      * <p>This method is a no-op if the index under the given name does not exist.
      *
+     * @param partitionId Partition ID.
      * @param indexId Index ID.
+     * @throws StorageException If the given partition does not exist, or fail destroying the index.
      */
-    CompletableFuture<Void> destroyIndex(UUID indexId);
+    CompletableFuture<Void> destroyIndex(int partitionId, UUID indexId) throws StorageException;
```
Comment (ibessonov) [inline modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/MvTableStorage.java:130]: Why did you add a partition id? What does it mean? I don't understand it.

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java [hunk 3]
```diff
@@ -311,51 +313,89 @@ public void testMisconfiguredIndices() {
...
-                    clock.now()
-            );
+            partitionStorage.addWriteCommitted(rowId, binaryRow, clock.now());
 
             partitionStorage.lastAppliedIndex(100);
 
             return null;
         });
 
+        hashIndexStorage.put(indexRow);
```
Comment (ibessonov) [inline modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java:340]: Why are you doing it without "runConsistently"? All writes should be inside of the closure.
And index updates should be in the same closure as writes, aren't they? It's irrelevant for this particular test, but as a general rule it's fine, why violating it?

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java [hunk 4]
```diff
@@ -311,51 +313,89 @@ public void testMisconfiguredIndices() {
...
         MvPartitionStorage newPartitionStorage0 = tableStorage.getMvPartition(PARTITION_ID);
+        HashIndexStorage newHashIndexStorage0 = tableStorage.getOrCreateHashIndex(PARTITION_ID, hashIndexId);
+        SortedIndexStorage newSortedIndexStorage0 = tableStorage.getOrCreateSortedIndex(PARTITION_ID, sortedIndexId);
 
         assertNotNull(newPartitionStorage0);
+        assertNotNull(newHashIndexStorage0);
+        assertNotNull(newSortedIndexStorage0);
+
         assertNotSame(partitionStorage, newPartitionStorage0);
+        assertNotSame(hashIndexStorage, newHashIndexStorage0);
```
Comment (ibessonov) [inline modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java:357]: Why do you assert things like these? They look completely wrong. What if we have a partition listener that has a cached value of MvPartitionStorage? Should it invalidate it and ask for another one?
Old instances must work. Please fix the test and the test implementation

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/impl/TestMvTableStorage.java [hunk 5]
```diff
@@ -40,50 +37,27 @@
...
+     * Storages container for the partition.
      */
-    private static class SortedIndices {
-        private final SortedIndexDescriptor descriptor;
-
-        final Map<Integer, SortedIndexStorage> storageByPartitionId = new ConcurrentHashMap<>();
-
-        SortedIndices(SortedIndexDescriptor descriptor) {
-            this.descriptor = descriptor;
+    private static class Storages {
```
Comment (ibessonov) [inline modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/impl/TestMvTableStorage.java:51]: Would it be better to store indexes in the "TestMvPartitionStorage" instance? What's the ultimate purpose of a separate class?

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/impl/TestMvTableStorage.java [hunk 6]
```diff
@@ -93,62 +67,59 @@ public TestMvTableStorage(TableConfiguration tableCfg, TablesConfiguration table
...
-            throw new StorageException("Partition ID " + partitionId + " does not exist");
-        }
+    public TestSortedIndexStorage getOrCreateSortedIndex(int partitionId, UUID indexId) throws StorageException {
+        Storages storages = storagesByPartitionId.get(partitionId);
 
-        SortedIndices sortedIndices = sortedIndicesById.computeIfAbsent(
-                indexId,
-                id -> new SortedIndices(new SortedIndexDescriptor(id, tablesCfg.value()))
-        );
+        checkPartitionStoragesExists(storages, partitionId);
```
Comment (ibessonov) [inline modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/impl/TestMvTableStorage.java:93]: Should we lazily create a partition? What does the documentation say? I believe that some of storages have different behavior, I don't like that

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/impl/TestMvTableStorage.java [hunk 7]
```diff
@@ -180,33 +151,51 @@ public void destroy() throws StorageException {
...
+
+        checkPartitionStoragesExists(oldStorages, partitionId);
 
-        assert oldPartitionStorage != null : "Partition does not exist: " + partitionId;
+        if (backupStoragesByPartitionId.putIfAbsent(partitionId, oldStorages) == null) {
+            Storages newStorages = new Storages(new TestMvPartitionStorage(partitionId));
 
-        if (backupPartitions.putIfAbsent(partitionId, oldPartitionStorage) == null) {
-            partitions.put(partitionId, new TestMvPartitionStorage(partitionId));
+            oldStorages.hashIndexes.keySet().forEach(indexId -> newStorages.hashIndexes.put(
```
Comment (ibessonov) [inline modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/impl/TestMvTableStorage.java:162]: Creating indexes is not mandatory, as far as I see. Rebalance itself will create indexes that it needs

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/MvTableStorage.java (hunk 1 above)
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/MvTableStorage.java:178]: Fix it

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java (hunk 3 above)
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java:340]: Technically it can do this, but you're right, let it be inside `runConsistently`.

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java (hunk 4 above)
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java:357]: Here we check that all storages are new and empty.
Why is this wrong?

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/impl/TestMvTableStorage.java (hunk 5 above)
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/impl/TestMvTableStorage.java:51]: It's just clearer and we change the existing code less, if you insist, I can move it to `TestMvPartitionStorage`.

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/impl/TestMvTableStorage.java (hunk 6 above)
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/impl/TestMvTableStorage.java:93]: Please see the documentation for the base method. =)
I did not change it and kept the logic.

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/impl/TestMvTableStorage.java (hunk 7 above)
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/impl/TestMvTableStorage.java:162]: I suggest discussing this in person.

Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/impl/TestMvTableStorage.java:162]: We decided personally that we would recreate them.

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/MvTableStorage.java (hunk 2 above)
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/MvTableStorage.java:130]: Fix it

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java (hunk 4 above)
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvTableStorageTest.java:357]: Fix it

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/impl/TestMvTableStorage.java (hunk 5 above)
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/impl/TestMvTableStorage.java:51]: Discussed to do it through delegates so that there are less problems when we change storages. 
Therefore, move `TestHashIndexStorage`, `TestSortedIndexStorage` to `TestMvPartitionStorage`.

Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/impl/TestMvTableStorage.java:51]: Fix it

Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/impl/TestMvTableStorage.java:61]: Tried to chage it.
